File: src/multivariable_calculus_partial_derivatives_hard.py
"""
multivariable_calculus - partial_derivatives (hard)
Generated: 2026-02-22T05:34:28.324036
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 5)

    if problem_type == 1:
        # Mixed partial f_xy
        func_type = choice(['poly', 'trig', 'exp'])
        if func_type == 'poly':
            a = nonzero(-4, 4)
            b = nonzero(-4, 4)
            c = nonzero(-4, 4)
            xp = randint(2, 4)
            yp = randint(2, 4)
            f = a * x**xp * y**yp + b * x**2 * y + c * x * y**2
        elif func_type == 'trig':
            a = nonzero(-3, 3)
            b = nonzero(2, 5)
            f = a * sin(b * x * y)
        else:
            a = nonzero(-2, 2)
            b = nonzero(1, 4)
            c = nonzero(1, 4)
            f = a * exp(b * x + c * y)
        f_xy = simplify(diff(diff(f, x), y))
        return problem(
            question=f"Find $\\frac{{\\partial^2 f}}{{\\partial x \\partial y}}$ for $f(x,y) = {latex(f)}$",
            answer=f_xy,
            difficulty=(1600, 1700),
            topic="multivariable_calculus/partial_derivatives",
            solution=steps(
                f"$\\frac{{\\partial f}}{{\\partial x}} = {latex(diff(f, x))}$",
                f"Differentiate with respect to $y$:",
                f"$\\frac{{\\partial^2 f}}{{\\partial x \\partial y}} = {latex(f_xy)}$"
            )
        )

    elif problem_type == 2:
        # Chain rule: f(u(t), v(t)), find df/dt
        a = nonzero(-3, 3)
        b = nonzero(-3, 3)
        m = randint(1, 3)
        n = randint(1, 3)
        u_func = m * t**2
        v_func = n * t
        f_expr = a * x**2 + b * y**2
        f_u_sub = diff(f_expr, x).subs(x, u_func)
        f_v_sub = diff(f_expr, y).subs(y, v_func)
        du_dt = diff(u_func, t)
        dv_dt = diff(v_func, t)
        ans = simplify(f_u_sub * du_dt + f_v_sub * dv_dt)
        return problem(
            question=f"Let $f(u,v) = {latex(a)}u^2 + {latex(b)}v^2$ where $u = {latex(u_func)}$, $v = {latex(v_func)}$. Find $\\frac{{df}}{{dt}}$.",
            answer=ans,
            difficulty=(1650, 1750),
            topic="multivariable_calculus/partial_derivatives",
            solution=steps(
                f"$\\frac{{\\partial f}}{{\\partial u}} = {latex(2*a)}u$, $\\frac{{\\partial f}}{{\\partial v}} = {latex(2*b)}v$",
                f"$\\frac{{du}}{{dt}} = {latex(du_dt)}$, $\\frac{{dv}}{{dt}} = {latex(dv_dt)}$",
                f"Chain rule: $\\frac{{df}}{{dt}} = \\frac{{\\partial f}}{{\\partial u}}\\frac{{du}}{{dt}} + \\frac{{\\partial f}}{{\\partial v}}\\frac{{dv}}{{dt}}$",
                f"$= {latex(ans)}$"
            )
        )

    elif problem_type == 3:
        # Directional derivative
        a = nonzero(-3, 3)
        b = nonzero(-3, 3)
        c = nonzero(1, 3)
        f = a * x**2 + b * y**2 + c * x * y
        x0 = randint(-2, 2)
        y0 = randint(-2, 2)
        v1 = choice([-3, -2, -1, 1, 2, 3])
        v2 = choice([-3, -2, -1, 1, 2, 3])
        v_mag = sqrt(v1**2 + v2**2)
        grad_x = diff(f, x).subs([(x, x0), (y, y0)])
        grad_y = diff(f, y).subs([(x, x0), (y, y0)])
        ans = simplify((grad_x * v1 + grad_y * v2) / v_mag)
        return problem(
            question=f"Find the directional derivative of $f(x,y) = {latex(f)}$ at $({x0},{y0})$ in direction $\\langle {v1}, {v2} \\rangle$.",
            answer=ans,
            difficulty=(1700, 1800),
            topic="multivariable_calculus/partial_derivatives",
            solution=steps(
                f"Gradient: $\\nabla f = \\langle {latex(diff(f,x))}, {latex(diff(f,y))} \\rangle$",
                f"At $({x0},{y0})$: $\\nabla f = \\langle {latex(grad_x)}, {latex(grad_y)} \\rangle$",
                f"Unit vector: $\\mathbf{{u}} = \\frac{{1}}{{{latex(v_mag)}}}\\langle {v1}, {v2} \\rangle$",
                f"$D_\\mathbf{{u}}f = \\nabla f \\cdot \\mathbf{{u}} = {latex(ans)}$"
            )
        )

    elif problem_type == 4:
        # Implicit differentiation via partial derivatives: F(x,y) = 0
        a = nonzero(-3, 3)
        b = nonzero(-3, 3)
        c = nonzero(1, 3)
        const = randint(1, 8)
        F = a * x**2 + b * y**2 + c * x * y - const
        F_x = diff(F, x)
        F_y = diff(F, y)
        ans = simplify(-F_x / F_y)
        return problem(
            question=f"Use implicit differentiation to find $\\frac{{dy}}{{dx}}$ for ${latex(F)} = 0$.",
            answer=ans,
            difficulty=(1650, 1750),
            topic="multivariable_calculus/partial_derivatives",
            solution=steps(
                f"$\\frac{{dy}}{{dx}} = -\\frac{{F_x}}{{F_y}}$",
                f"$F_x = {latex(F_x)}$",
                f"$F_y = {latex(F_y)}$",
                f"$\\frac{{dy}}{{dx}} = {latex(ans)}$"
            )
        )

    else:
        # Higher-order mixed partials
        a = nonzero(-3, 3)
        b = nonzero(-3, 3)
        c = nonzero(1, 2)
        xp = randint(3, 4)
        yp = randint(3, 4)
        f = a * x**xp * y**yp + b * x * y**3 + c * x**2 * y**2
        order_type = choice(['xxy', 'xyy', 'xyx'])
        if order_type == 'xxy':
            ans = simplify(diff(diff(diff(f, x), x), y))
            notation = "\\frac{\\partial^3 f}{\\partial y \\partial x^2}"
        elif order_type == 'xyy':
            ans = simplify(diff(diff(diff(f, x), y), y))
            notation = "\\frac{\\partial^3 f}{\\partial y^2 \\partial x}"
        else:
            ans = simplify(diff(diff(diff(f, x), y), x))
            notation = "\\frac{\\partial^3 f}{\\partial x \\partial y \\partial x}"
        return problem(
            question=f"Find ${notation}$ for $f(x,y) = {latex(f)}$",
            answer=ans,
            difficulty=(1750, 1850),
            topic="multivariable_calculus/partial_derivatives",
            solution=steps(
                f"$\\frac{{\\partial f}}{{\\partial x}} = {latex(diff(f, x))}$",
                f"Continue differentiating step by step",
                f"${notation} = {latex(ans)}$"
            )
        )

emit(generate())

File: src/multivariable_calculus_partial_derivatives_hard.jl
# multivariable_calculus - partial_derivatives (hard)
# Generated: 2026-03-08T20:58:26.595617

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x y z begin
    set_topic!("multivariable_calculus/partial_derivatives")
    
    problem_type = rand(1:6)
    
    if problem_type == 1
        # Chain rule with partial derivatives
        a = nonzero(-8, 8)
        b = nonzero(-8, 8)
        c = nonzero(-8, 8)
        d = nonzero(-8, 8)
        m = rand(2:4)
        n = rand(2:4)
        
        f_expr = a*x^m + b*y^n
        x_of_t = c*z^2 + d
        y_of_t = nonzero(-8, 8)*z
        
        dfdx = diff(f_expr, x)
        dfdy = diff(f_expr, y)
        dxdz = diff(x_of_t, z)
        dydz = diff(y_of_t, z)
        
        dfdt_val = expand(dfdx * dxdz + dfdy * dydz)
        
        problem(
            question="Let \$f(x,y) = $(tex(f_expr))\$ where \$x = $(tex(x_of_t))\$ and \$y = $(tex(y_of_t))\$. Find \$\\frac{df}{dz}\$ using the chain rule.",
            answer=dfdt_val,
            difficulty=(1800, 2000),
            solution=steps(
                sol("f(x,y)", f_expr),
                sol("\\frac{\\partial f}{\\partial x}", dfdx),
                sol("\\frac{\\partial f}{\\partial y}", dfdy),
                sol("\\frac{dx}{dz}", dxdz),
                sol("\\frac{dy}{dz}", dydz),
                "Apply chain rule: \$\\frac{df}{dz} = \\frac{\\partial f}{\\partial x}\\frac{dx}{dz} + \\frac{\\partial f}{\\partial y}\\frac{dy}{dz}\$",
                sol("Answer", dfdt_val)
            ),
            time=180
        )
        
    elseif problem_type == 2
        # Second-order mixed partial derivatives (Clairaut's theorem)
        a = nonzero(-6, 6)
        b = nonzero(-6, 6)
        c = nonzero(-6, 6)
        m = rand(2:3)
        n = rand(2:3)
        
        f_expr = a*x^m*y^n + b*x*y^2 + c*y
        
        fxy = diff(diff(f_expr, x), y)
        fyx = diff(diff(f_expr, y), x)
        
        problem(
            question="For \$f(x,y) = $(tex(f_expr))\$, compute \$f_{xy}\$ and verify that \$f_{xy} = f_{yx}\$.",
            answer=fxy,
            difficulty=(1800, 2100),
            solution=steps(
                sol("f(x,y)", f_expr),
                sol("f_x", diff(f_expr, x)),
                sol("f_{xy}", fxy),
                sol("f_y", diff(f_expr, y)),
                sol("f_{yx}", fyx),
                "By Clairaut's theorem, \$f_{xy} = f_{yx}\$ (verified)",
                sol("Answer", fxy)
            ),
            time=240
        )
        
    elseif problem_type == 3
        # Directional derivative
        a = nonzero(-8, 8)
        b = nonzero(-8, 8)
        c = nonzero(-8, 8)
        x0 = randint(-5, 5)
        y0 = randint(-5, 5)
        
        vx = nonzero(-4, 4)
        vy = nonzero(-4, 4)
        v_mag = sqrt(vx^2 + vy^2)
        
        f_expr = a*x^2 + b*x*y + c*y^2
        
        fx = diff(f_expr, x)
        fy = diff(f_expr, y)
        
        fx_val = substitute(fx, Dict([x => x0, y => y0]))
        fy_val = substitute(fy, Dict([x => x0, y => y0]))
        
        D_u = (fx_val * vx + fy_val * vy) / v_mag
        
        problem(
            question="Find the directional derivative of \$f(x,y) = $(tex(f_expr))\$ at the point \$($(x0), $(y0))\$ in the direction of the vector \$\\langle $(vx), $(vy) \\rangle\$.",
            answer=D_u,
            difficulty=(2000, 2300),
            solution=steps(
                sol("f(x,y)", f_expr),
                sol("\\nabla f", "\\langle $(tex(fx)), $(tex(fy)) \\rangle"),
                sol("\\nabla f($(x0), $(y0))", "\\langle $(fx_val), $(fy_val) \\rangle"),
                "Unit vector: \$\\mathbf{u} = \\frac{1}{$(tex(v_mag))}\\langle $(vx), $(vy) \\rangle\$",
                "Directional derivative: \$D_{\\mathbf{u}}f = \\nabla f \\cdot \\mathbf{u}\$",
                sol("Answer", D_u)
            ),
            time=210
        )
        
    elseif problem_type == 4
        # Tangent plane equation
        a = nonzero(-6, 6)
        b = nonzero(-6, 6)
        c = nonzero(-6, 6)
        x0 = randint(-3, 3)
        y0 = randint(-3, 3)
        
        f_expr = a*x^2 + b*y^2 + c*x*y
        
        fx = diff(f_expr, x)
        fy = diff(f_expr, y)
        
        z0 = substitute(f_expr, Dict([x => x0, y => y0]))
        fx_val = substitute(fx, Dict([x => x0, y => y0]))
        fy_val = substitute(fy, Dict([x => x0, y => y0]))
        
        tangent = z0 + fx_val*(x - x0) + fy_val*(y - y0)
        
        problem(
            question="Find the equation of the tangent plane to \$z = $(tex(f_expr))\$ at the point \$($(x0), $(y0), $(z0))\$.",
            answer=expand(tangent),
            difficulty=(1900, 2200),
            solution=steps(
                sol("z", f_expr),
                sol("\\frac{\\partial z}{\\partial x}", fx),
                sol("\\frac{\\partial z}{\\partial y}", fy),
                "At \$($(x0), $(y0))\$: \$\\frac{\\partial z}{\\partial x} = $(fx_val)\$, \$\\frac{\\partial z}{\\partial y} = $(fy_val)\$",
                "Tangent plane: \$z - $(z0) = $(fx_val)(x - $(x0)) + $(fy_val)(y - $(y0))\$",
                sol("Answer", expand(tangent))
            ),
            time=240
        )
        
    elseif problem_type == 5
        # Implicit differentiation for partial derivatives
        a = nonzero(-5, 5)
        b = nonzero(-5, 5)
        c = nonzero(-5, 5)
        
        F_expr = a*x^2 + b*y^2 + c*x*y*z
        
        Fx = diff(F_expr, x)
        Fz = diff(F_expr, z)
        
        dzdx = -Fx / Fz
        
        problem(
            question="Given the implicit equation \$$(tex(F_expr)) = 0\$, find \$\\frac{\\partial z}{\\partial x}\$ using implicit differentiation.",
            answer=simplify(dzdx),
            difficulty=(2100, 2400),
            solution=steps(
                sol("F(x,y,z)", F_expr ~ 0),
                "Differentiate implicitly with respect to \$x\$",
                sol("F_x", Fx),
                sol("F_z", Fz),
                "By implicit function theorem: \$\\frac{\\partial z}{\\partial x} = -\\frac{F_x}{F_z}\$",
                sol("Answer", simplify(dzdx))
            ),
            time=270
        )
        
    else
        # Critical points and second derivative test
        a = choice([-2, -1, 1, 2])
        b = choice([-2, -1, 1, 2])
        c = randint(-8, 8)
        d = randint(-8, 8)
        
        f_expr = a*x^2 + b*y^2 + c*x + d*y
        
        fx = diff(f_expr, x)
        fy = diff(f_expr, y)
        
        crit_x = -c // (2*a)
        crit_y = -d // (2*b)
        
        fxx = diff(fx, x)
        fyy = diff(fy, y)
        fxy = diff(fx, y)
        
        D_val = fxx * fyy - fxy^2
        
        if D_val > 0 && a > 0
            classification = "local minimum"
        elseif D_val > 0 && a < 0
            classification = "local maximum"
        else
            classification = "saddle point"
        end
        
        ans_point = "$(fmt_tuple([crit_x, crit_y]))"
        
        problem(
            question="Find and classify all critical points of \$f(x,y) = $(tex(f_expr))\$. Give your answer as an ordered pair.",
            answer=ans_point,
            difficulty=(2000, 2400),
            solution=steps(
                sol("f(x,y)", f_expr),
                sol("f_x", fx),
                sol("f_y", fy),
                "Set \$f_x = 0\$ and \$f_y = 0\$",
                sol("Critical point", ans_point),
                "Second derivative test: \$f_{xx} = $(fxx)\$, \$f_{yy} = $(fyy)\$, \$f_{xy} = $(fxy)\$",
                "Discriminant: \$D = f_{xx}f_{yy} - (f_{xy})^2 = $(D_val)\$",
                "Classification: $(classification)",
                sol("Answer", ans_point)
            ),
            time=300
        )
    end
end